feat: add `ndarray/flatten`

headlessNode opened the pull request:
---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes. report:
  - task: lint_filenames status: passed
  - task: lint_editorconfig status: passed
  - task: lint_markdown status: passed
  - task: lint_package_json status: passed
  - task: lint_repl_help status: passed
  - task: lint_javascript_src status: passed
  - task: lint_javascript_cli status: na
  - task: lint_javascript_examples status: passed
  - task: lint_javascript_tests status: passed
  - task: lint_javascript_benchmarks status: passed
  - task: lint_python status: na
  - task: lint_r status: na
  - task: lint_c_src status: na
  - task: lint_c_examples status: na
  - task: lint_c_benchmarks status: na
  - task: lint_c_tests_fixtures status: na
  - task: lint_shell status: na
  - task: lint_typescript_declarations status: passed
  - task: lint_typescript_tests status: passed
  - task: lint_license_headers status: passed ---

Resolves https://github.com/stdlib-js/metr-issue-tracker/issues/75.

## Description

> What is the purpose of this pull request?

This pull request:

-   add `ndarray/flatten`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/75
-   progresses #2656

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add ndarra/flatten

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
fix: import in benchmark file

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/README.md:
````diff
@@ -0,0 +1,127 @@
+var flatten = require( '@stdlib/ndarray/flatten' );
+```
+
+#### flatten( x, depth )
````
```suggestion
#### flatten( x\[, depth] )
```

This is a top-level package. We should default to flattening to a one-dimensional ndarray. This applies here and everywhere.

kgryte replied:
i.e., by default `depth` should be `infinity`.

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/README.md:
```diff
@@ -0,0 +1,127 @@
+The function accepts the following arguments:
+
+-   **x**: input [`ndarray`][@stdlib/ndarray/ctor].
+-   **depth**: a non-negative integer specifying the number of input [`ndarray`][@stdlib/ndarray/ctor] dimensions to flatten.
```
```suggestion
-   **depth**: a nonnegative integer specifying the number of input [`ndarray`][@stdlib/ndarray/ctor] dimensions to flatten.
```

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/docs/repl.txt:
```diff
@@ -0,0 +1,30 @@
+
+    depth: integer
+        A non-negative integer specifying the number of input ndarray dimensions
+        to flatten.
```
```suggestion
        Number of dimensions to flatten.
```

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/docs/repl.txt:
```diff
@@ -0,0 +1,30 @@
+    Returns a flattened copy of an input ndarray.
+
+    The `depth` must be a valid non-negative integer.
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/docs/repl.txt:
```diff
@@ -0,0 +1,30 @@
+
+{{alias}}( x, depth )
```
```suggestion
{{alias}}( x[, depth] )
```

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/README.md:
````diff
@@ -0,0 +1,127 @@
+var flatten = require( '@stdlib/ndarray/flatten' );
+```
+
+#### flatten( x, depth )
````
Also, similar to NumPy (ref: https://numpy.org/doc/stable/reference/generated/numpy.ravel.html#numpy-ravel), we should support specifying a "read" order.

```
flatten( x[, opts] )
```

with options

- **depth**: default infinity.
- **order**: read order. If `row-major`, always flatten an input array in lexicographic order. If `column-major`, always flatten in colexicographic order. If `same`, flatten in the order of the input array (i.e., equivalent to `order = getOrder( x )`). If `any`, flatten in colexicographic order if input array **data** is column-major contiguous; otherwise, flatten in lexicographic order (i.e., flatten according to how the data is arranged in memory, which may not correspond to the stated array order).

kgryte replied:
Additional references:

- Order options: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/ndarray/array
- "colexicographic" terminology: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/array/base/flatten

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/README.md:
```diff
@@ -0,0 +1,127 @@
+
+## Notes
+
+-   The provided input [`ndarray`][@stdlib/ndarray/ctor] **must** have more than one-dimension.
```
I don't think this is correct. Flattening a one-dimensional ndarray should be a no-op, but, as we are always returning a new array, providing a one-dimensional array should just be a straightforward copy.

kgryte replied:
The more interesting question is whether we should support zero-dimensional ndarrays. I suggest seeing what NumPy allows.

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+	}
+	d = ndims;
+	if ( isPositiveInteger( depth ) ) {
+		if ( depth > ndims - 1 ) {
```
Why should this be an error condition? Why not just clamp to the number of dimensions? I am not sure how likely this is, but I suppose a user could want to flatten a list of ndarrays of varying ranks and want to flatten them to at most a specified depth. By raising an exception here, you disallow this kind of usage.

Think of this sort of clamping behavior as being similar to `indexOf` and friends in the handling of out-of-bounds values.

kgryte replied:
Similar to https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/flatten-array, `depth` should be considered the "maximum depth" to flatten, rather than a strict edict.

You should also always sanity check against built-in behavior:

```
In [29]: [[1,2,3]].flat()
Out[29]: [ 1, 2, 3 ]

In [30]: [[1,2,3]].flat(1)
Out[30]: [ 1, 2, 3 ]

In [31]: [[1,2,3]].flat(10)
Out[31]: [ 1, 2, 3 ]
```

Notice the behavior in the last example.

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+		throw new RangeError( format( 'invalid argument. First argument be an ndarray-like object with more than one dimension. Value: `%s`.', ndims ) );
+	}
+	d = ndims;
+	if ( isPositiveInteger( depth ) ) {
```
Why a positive integer? Why not a nonnegative integer? If `0`, we don't flatten.

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+* Returns a flattened copy of an input ndarray.
+*
+* @param {ndarray} x - input ndarray
+* @param {integer} depth - number of dimensions to flatten
```
This should be `NonNegativeInteger`

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,100 @@
+	}
+	osh = xsh.slice( d );
+	osh.unshift( s );
+	return new ndarray( getDtype( x ), getData( x ), osh, shape2strides( osh, getOrder( x ) ), 0, getOrder( x ) ); // eslint-disable-line max-len
```
This is very clearly not correct for a number of reasons:

1. You are returning a view here.
2. This doesn't work for non-contiguous arrays.

Consider a 3x2 ndarray having the following data buffer

```
[ 1, 2, X, X, 3, 4, X, X, 5, 6, X, X ]
```

where `X` entries correspond to non-indexed elements in the buffer. In this example, for a row-major ndarray, the strides would be `[ 4, 1 ]`.

Suppose we flatten to a one-dimensional ndarray having shape `[ 6 ]`. According to your logic, the strides should be `[ 1 ]`, but that is very clearly incorrect. There isn't a stride that allows iterating over elements contiguously.

You need to spend more time thinking through this implementation. As mentioned elsewhere, only under special circumstances can one avoid explicit data copies and that is when ndarray data is contiguous.

Here, we should always be performing a copy, as conditional copy-view semantics often leads to undesired behavior (e.g., later writing into an ndarray which shares memory with one or more other ndarrays leading to undesired mutation). There can be instances where someone wants to create a flattened view, but we should create a separate package for that.

kgryte reviewed (changes requested):
Left an initial review. This implementation needs some significant surgery before it can move forward.

headlessNode committed:
refactor: most of the stuff

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/README.md:
```diff
@@ -0,0 +1,174 @@
+
+-   **depth**: a nonnegative integer specifying the number of input [`ndarray`][@stdlib/ndarray/ctor] dimensions to flatten.
+
+-   **order**: order of the output [`ndarray`][@stdlib/ndarray/ctor]. The option may be one of the following values:
```
```suggestion
-   **order**: order of the output [`ndarray`][@stdlib/ndarray/ctor]. May be one of the following:
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/README.md:
```diff
@@ -0,0 +1,174 @@
+    -   `'any'`: the input [`ndarray`][@stdlib/ndarray/ctor] is flattened in colexicographic [order][@stdlib/ndarray/orders] if the input [`ndarray`][@stdlib/ndarray/ctor] is column-major; otherwise, it is flattened in lexicographic [order][@stdlib/ndarray/orders].
+    -   `'same'`: the input [`ndarray`][@stdlib/ndarray/ctor] is flattened in the [order][@stdlib/ndarray/orders] of the input [`ndarray`][@stdlib/ndarray/ctor].
+
+By default, all the dimensinos of the input [ndarray][@stdlib/ndarray/ctor] are flattened. To flatten upto to a specific `depth`, specify the `depth` option.
```
```suggestion
By default, all the dimensions of the input [ndarray][@stdlib/ndarray/ctor] are flattened. To flatten to a desired depth, specify the `depth` option.
```

kgryte committed:
docs: update copy

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,152 @@
+					} else {
+						opts.order = getOrder( x );
+					}
+				}
+				else if ( options.order === 'same' ) {
```
```suggestion
				} else if ( options.order === 'same' ) {
```

kgryte committed:
style: remove linebreak

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,151 @@
+* @param {ndarray} x - input ndarray
+* @param {Object} [options] - function options
+* @param {NonNegativeInteger} [options.depth] - number of dimensions to flatten
+* @param {String} [options.order] - order of the output ndarray
```
```suggestion
* @param {string} [options.order='row-major'] - order in which input ndarray elements are flattened
```

kgryte committed:
docs: fix description

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/ndarray/flatten/lib/main.js:
```diff
@@ -0,0 +1,151 @@
+* @param {string} [options.order='row-major'] - order in which input ndarray elements are flattened
+* @throws {TypeError} first argument must be an ndarray-like object
+* @throws {TypeError} options argument must be an object
+* @throws {TypeError} options.depth must be a nonnegative integer
+* @throws {TypeError} options.order must be a recognized order
```
```suggestion
* @throws {TypeError} must provide valid options
```

kgryte committed:
docs: update description based on conventions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
Merge branch 'develop' of https://github.com/stdlib-js/stdlib into pr/headlessNode/8021

kgryte committed:
chore: clean-up

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

kgryte committed:
test: add and update tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte reviewed (approved):
LGTM after clean-up.
